### DISCUSSION TITLE
Custom Client Implications Questions

### DISCUSSION THREAD

**Discussion by wchest** [mozilla/hubs#3770] [Hubs Cloud Questions] [answered]
We are considering following the steps listed here: to create a custom client on top of Hubs and I have a few quick questions.

1. When we set up the fork and add this repo as the upstream, will this impact the automatic upgrades we currently receive? As I understand it, as a Hubs Cloud customer you all push out periodic updates behind the scenes to all Hubs Cloud instances. I know on some occasions, the bump from 1.0.0 to 1.1.0 you all recommended a manual upgrade (if the automatic upgrade failed), but we've had our Hubs Cloud instance for over 8 months now and it seems like we're getting automatic updates. Will we need to now pull these changes and update our stack manually if we fork? If so, how do we know when to make these updates?
2. One reason to potentially create a custom client is that we'd like to be able to use custom authentication for rooms. We often send people who have been authenticated already in our system to rooms in our Hubs Cloud instance, but we currently lack any way to prevent non-authenticated users (on our side) from accessing rooms. This has been a challenge for many of our partners who want to make sure access is controlled in some way (ie no one can just share the link to a room on Twitter). I've been following the discussions here: https://github.com/mozilla/hubs-cloud/issues/27 and it sounds like there have been some proposals made around allowing us to programmatically create identities tied to email accounts, but it sounds like even in this case the user would still need to log in once they arrived in Hubs. Either way, two questions come to mind:
 * Has there been any progress on this piece?
 * If we wanted to implement some sort of auth check, my understanding is doing it in a fork of the client wouldn't be sufficient from a security perspective (although it may be enough to satisfy some of our partners). If we had to do a more comprehensive authentication piece, would we need to also fork Reticiulum? Is there any precedent for this and do you know of any other groups that have tried this? I would obviously like to avoid forking Reticulum and the client if possible.
 
Thank you for all you do to create this software, we're really loving using it for our work!

**Comment by robinkwilson** [comment, marked as answer]
1. If you deploy a custom client, your backend will still update automatically and you will be responsible for merging in client updates. We're working on a better process for letting the community know of these updates. We're working on getting a newsletter/email list for these updates. - Once that gets finalized I'll link the sign up list here.

Right now if you join our discord community we have a role "Hubs Client Devs" we can add you to. Click on "Community" on this page: https://hubs.mozilla.com/ and DM me, I'll get you the role. 

2. Auth
   - We have some progress, but nothing finalized or moving forward. 
   - Unfortunately yes, securing on the backend is best for security. 
      - I believe you already know about having the two auth systems working in tandem described here in the docs: https://hubs.mozilla.com/docs/hubs-cloud-limiting-user-access.html. First you'd have them sign up in your auth system, then make an account for them in the hubs system so they can sign in there too. (Unfortunately the user then signs in twice for both systems)
      - If you're using subdomains, I've been able to secure the front end with a custom client for each page by sending a request to the auth system using a secured cookie on the subdomain and checking against that. Then allowing the room to load, if signed in, or redirecting the user to the sign in page in the other system, if not signed in.

I'd recommend against forking reticulum, there's no recommended way to do this or get it running locally right now. 

We're always happy to hear people are enjoying using hubs!

**Reply by robinkwilson** [reply]
Marking as answered for now, let us know if you have any additional specific questions then "unmark" to get it back in the pipeline!
